solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -198,6 +200,12 @@
 
   private volatile SortedSet<String> liveNodes = emptySortedSet();
 
+  private volatile SortedSet<String> liveQueryNodes = emptySortedSet();
+
+  private final Object refreshLiveQueryNodesLock = new Object();
+
+  private final AtomicReference<SortedSet<String>> lastFetchedLiveQueryNodes = new AtomicReference<>();
+
```

Comment:
I was wondering where the lock and last fetched fields for `live_nodes` were. I see they are way down on line `828`. Nothing for you to change. But I'm just noting that I wish the Solr code base was more consistent. Keeping the mutexes up here where you have put them is what our style guide would have recommended.